File: Sliding Window Algorithms/LeetCode 0438.py
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        n, m = len(s), len(p)
        if m > n:
            return []

        scount = [0] * 26
        pcount = [0] * 26
        ans = []

        for c in p:
            pcount[ord(c) - ord('a')] += 1

        for i in range(n):
            scount[ord(s[i]) - ord('a')] += 1
            if i >= m:
                scount[ord(s[i - m]) - ord('a')] -= 1
            if scount == pcount:
                ans.append(i - m + 1)

        return ans

File: Sliding Window Algorithms/LeetCode 0438.java
import java.util.*;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int n = s.length(), m = p.length();
        List<Integer> ans = new ArrayList<>();
        if(m > n)
        {
            return ans;
        }
        int[] scount = new int[26];
        int[] pcount = new int[26];

        for(char c : p.toCharArray())
        {
            pcount[c - 'a']++;
        }
        for(int i = 0; i < n; i++)
        {
            scount[s.charAt(i) - 'a']++;
            if(i >= m)
            {
                scount[s.charAt(i - m) - 'a']--;
            }
            if(Arrays.equals(scount, pcount))
            {
                ans.add(i - m + 1);
            }
        }
        return ans;
    }
}
